### PR TITLE
tox4

### PR DESCRIPTION
**Pull request by frostming**

### CODE REVIEW

File: tox_pdm/plugin4.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    from tox.execute.api import ExecuteInstance, ExecuteOptions
+    from tox.execute.request import ExecuteRequest
+    from tox.execute.stream import SyncWrite
+    from tox.tox_env.api import ToxEnvCreateArgs
+    from tox.tox_env.register import ToxEnvRegister
+
+
+@impl
+def tox_add_option(parser: ArgumentParser) -> None:
+    os.environ["TOX_TESTENV_PASSENV"] = "PYTHONPATH"
```
Comment (gaborbernat) [inline tox_pdm/plugin4.py:38]: This is hacky and will likely mean users can no longer use this. How come you need to set this? Feels against the spirit of tox. If you must though extend this list https://github.com/tox-dev/tox/blob/rewrite/src/tox/tox_env/api.py#L176 (via subclass, call super and appending to it).

File: tox_pdm/plugin4.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    @property
+    def installer(self) -> Pip:
+        if self._installer is None:
+            self._installer = PipPep582(self)
+        return self._installer
+
+    @property
+    def package_path(self) -> Path:
+        if not self._package_path:
+            self._package_path = Path(
```
Comment (gaborbernat) [inline tox_pdm/plugin4.py:71]: you likely can save here a line if you assign the line below to a local variable first

File: tox_pdm/plugin4.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+                get_env_lib_path(self.options.pdm, self.env_dir)
+            ).parent
+        return self._package_path
+
+    @property
+    def runs_on_platform(self) -> str:
+        return sys.platform
+
+    def create_python_env(self) -> None:
+        clone_pdm_files(self.env_dir, self.core["toxinidir"])
```
Comment (gaborbernat) [inline tox_pdm/plugin4.py:81]: you likely want here tox_root instead 👍

File: tox_pdm/plugin4.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+            extra={"executable": Path(interpreter.system_executable)},
+        )
+
+    def python_cache(self) -> t.Dict[str, t.Any]:
+        base = super().python_cache()
+        base.update({"executable": str(self.base_python.extra["executable"])})
+        return base
+
+    def env_bin_dir(self) -> Path:
+        return self.package_path / ("Scripts" if os.name == "nt" else "bin")
```
Comment (gaborbernat) [inline tox_pdm/plugin4.py:107]: Is this hard coded for PDM? Shouldn't you here follow what sysconfig says instead?

File: tox_pdm/plugin4.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+    def env_python(self) -> Path:
+        return t.cast(Path, self.base_python.extra["executable"])
+
+    def env_site_package_dir(self) -> Path:
+        return self.package_path / "lib"
```
Comment (gaborbernat) [inline tox_pdm/plugin4.py:113]: Is this hardcoded for pdm, same for both Windows and Linux?

File: tox_pdm/plugin4.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+class PdmRunner(Pdm, PythonRun):
+    def _setup_env(self) -> None:
+        super()._setup_env()
+        groups = self.conf["groups"]
+        cmd = ["pdm", "install", "--no-self"]
+        for group in groups:
+            cmd.extend(("--group", group))
+        self.execute(cmd, StdinSource.OFF)
+        bin_path = self.env_site_package_dir() / "bin"
```
Comment (gaborbernat) [inline tox_pdm/plugin4.py:127]: Always bin? Not scripts on Windows?

File: tox_pdm/plugin4.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+
+class PdmRunner(Pdm, PythonRun):
+    def _setup_env(self) -> None:
+        super()._setup_env()
+        groups = self.conf["groups"]
+        cmd = ["pdm", "install", "--no-self"]
+        for group in groups:
+            cmd.extend(("--group", group))
+        self.execute(cmd, StdinSource.OFF)
```
Comment (gaborbernat) [inline tox_pdm/plugin4.py:126]: Does this need to run all the time? Shouldn't it be cached on groups argument?

File: tox_pdm/plugin4.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+            desc="directory where to put project packages",
+        )
+
+    @property
+    def pkg_dir(self) -> Path:
+        return t.cast(Path, self.conf["pkg_dir"])
+
+    def perform_packaging(self, for_env: EnvConfigSet) -> t.List[Package]:
+        of_type: str = for_env["package"]
+        cmd = [
```
Comment (gaborbernat) [inline tox_pdm/plugin4.py:168]: Does this create a PEP-517 env all the time? Or does PDM under the hod caches the isolated build env?

File: tox_pdm/plugin4.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+        err: SyncWrite,
+    ) -> "ExecuteInstance":
+        return PDMRunExecuteInstance(request, options, out, err)
+
+
+class PDMRunExecuteInstance(LocalSubProcessExecuteInstance):
+    @property
+    def cmd(self) -> t.Sequence[str]:
+        if self._cmd is None:
+            pdm_exe = self.options._env.options.pdm
```
Comment (gaborbernat) [inline tox_pdm/plugin4.py:220]: You're accessing here a private field 🤔 This can break without warning.

File: tox_pdm/plugin4.py (hunk 6 above)
Comment (frostming) [reply tox_pdm/plugin4.py:127]: It is the pattern of `pip install --target`

File: tox_pdm/plugin4.py (hunk 7 above)
Comment (frostming) [reply tox_pdm/plugin4.py:126]: Good point, will improve

File: tox_pdm/plugin4.py (hunk 5 above)
Comment (frostming) [reply tox_pdm/plugin4.py:113]: Yes, it is hardcoded, sysconfig doesn't provide the ability to add a custom scheme.

Comment (gaborbernat) [reply tox_pdm/plugin4.py:113]: You mean, just for PDM (but not Python in general, which does)?

File: tox_pdm/plugin4.py (hunk 6 above)
Comment (gaborbernat) [reply tox_pdm/plugin4.py:127]: Can you expand on this?

File: tox_pdm/plugin4.py (hunk 5 above)
Comment (frostming) [reply tox_pdm/plugin4.py:113]: PDM uses a scheme that is different from any included in `sysconfig._INSTALL_SCHEME`, so I can't get the correct path by `sysconfig`'s provided methods.

Comment (gaborbernat) [reply tox_pdm/plugin4.py:113]: Makes sense 👍

File: tox_pdm/plugin4.py (hunk 6 above)
Comment (frostming) [reply tox_pdm/plugin4.py:127]: If you do `pip install --target mypath django`, the scripts will be created under `mypath/bin`, on Windows and POSIX systems.

File: tox_pdm/plugin4.py (hunk 9 above)
Comment (gaborbernat) [reply tox_pdm/plugin4.py:220]: The solution here might be that `options` needs to be passed for the execution request (in core).

File: tox_pdm/plugin4.py (hunk 6 above)
Comment (gaborbernat) [reply tox_pdm/plugin4.py:127]: Did not know, thanks for clarifying. Perhaps would be valuable to add this as a comment to the code.

**Pull request merged**